File: internal/infra/disk/disk.go
package disk

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"

	"mvmctl/internal/infra/errs"
)

// ── Disk-specific constants (moved from infra/constants.go) ──
const (
	MinRootSizeMB    = 500
	SizeTooSmallMB   = 100
	SectorSizeBytes  = 512
	MebibyteBytes    = 1024 * 1024
)

// DetectorWeights holds scoring weights for partition root detection.
var DetectorWeights = map[string]float64{
	"type_code":  1.0,
	"label":      0.8,
	"size":       0.5,
	"filesystem": 0.7,
}

// DetectorScores holds scoring values for partition root detection.
var DetectorScores = map[string]float64{
	"root_score":             1.0,
	"exclude_score":         -1.0,
	"neutral_score":          0.0,
	"mbr_linux_score":       0.5,
	"label_root_score":      1.0,
	"label_exclude_score":  -0.5,
	"size_largest_score":    0.5,
	"size_root_score":       0.3,
	"size_too_small_score": -0.5,
}

// ── Size multipliers (IEC binary units) ──

var sizeMultipliers = map[string]int64{
	"B":  1,
	"K":  1024,
	"KB": 1024,
	"M":  1024 * 1024,
	"MB": 1024 * 1024,
	"G":  1024 * 1024 * 1024,
	"GB": 1024 * 1024 * 1024,
	"T":  1024 * 1024 * 1024 * 1024,
	"TB": 1024 * 1024 * 1024 * 1024,
}

// output units for FormatDiskSize — built from sizeMultipliers map, excluding short forms.
// Mirrors Python's sorted(_SIZE_MULTIPLIERS.items(), key=lambda x: x[1], reverse=True) with skip of B/KB/MB/GB/TB.
// Initialised lazily via getFormatDiskUnits().
// TODO: call InitFormatDiskUnits() from app/app.go explicitly
var (
	formatDiskUnits []struct {
		suffix string
		size   int64
	}
	formatDiskUnitsOnce sync.Once
)

// getFormatDiskUnits builds and caches the sorted formatDiskUnits slice on first call.
func getFormatDiskUnits() []struct {
	suffix string
	size   int64
} {
	formatDiskUnitsOnce.Do(func() {
		shortForms := map[string]bool{"B": true, "KB": true, "MB": true, "GB": true, "TB": true}
		for suffix, size := range sizeMultipliers {
			if !shortForms[suffix] {
				formatDiskUnits = append(formatDiskUnits, struct {
					suffix string
					size   int64
				}{suffix, size})
			}
		}
		sort.Slice(formatDiskUnits, func(i, j int) bool {
			return formatDiskUnits[i].size > formatDiskUnits[j].size
		})
	})
	return formatDiskUnits
}

var sizePattern = regexp.MustCompile(`^(\d+(?:\.\d+)?)\s*([KMGT]?B?|[kmgt]?b?)?$`)

// ParseDiskSizeToBytes parses a disk size string like "512M", "1G", "2.5GB", "1024" into bytes.
// Matches Python's DiskUtils.parse_disk_size_to_bytes() exactly, including error messages.
func ParseDiskSizeToBytes(s string) (int64, error) {
	return ParseDiskSize(s)
}

// ParseDiskSize parses a disk size string into bytes.
// Matches Python's DiskUtils.parse_disk_size_to_bytes() exactly.
// Python raises MVMError on format errs.
func ParseDiskSize(s string) (int64, error) {
	s = strings.TrimSpace(s)
	upper := strings.ToUpper(s)
	match := sizePattern.FindStringSubmatch(upper)
	if match == nil {
		return 0, &errs.DomainError{
			Code:    errs.CodeValidationFailed,
			Class:   errs.ClassValidation,
			Message: fmt.Sprintf("Invalid disk size format: '%s'. Expected format: <number><unit> where unit is B, K, KB, M, MB, G, GB, T, TB", upper),
		}
	}

	numberStr, unit := match[1], match[2]
	if unit == "" {
		unit = "B" // Default to bytes if no unit
	}

	number, err := strconv.ParseFloat(numberStr, 64)
	if err != nil {
		return 0, &errs.DomainError{
			Code:    errs.CodeValidationFailed,
			Class:   errs.ClassValidation,
			Message: fmt.Sprintf("Invalid number in disk size: '%s'", numberStr),
		}
	}

	multiplier, ok := sizeMultipliers[unit]
	if !ok {
		return 0, &errs.DomainError{
			Code:    errs.CodeValidationFailed,
			Class:   errs.ClassValidation,
			Message: fmt.Sprintf("Unknown size unit: '%s'. Valid: B, K, KB, M, MB, G, GB, T, TB", unit),
		}
	}

	bytesCount := int64(number * float64(multiplier))
	if bytesCount < 0 {
		return 0, &errs.DomainError{
			Code:    errs.CodeValidationFailed,
			Class:   errs.ClassValidation,
			Message: fmt.Sprintf("Disk size cannot be negative: %s", upper),
		}
	}

	return bytesCount, nil
}

func FormatSectorsHumanReadable(sizeSectors int64, sectorSize int64) string {
	if sectorSize == 0 {
		sectorSize = 512
	}
	sizeBytes := sizeSectors * sectorSize
	sizeMib := float64(sizeBytes) / (1024 * 1024)
	if sizeMib >= 1024 {
		return fmt.Sprintf("%.1f GiB", sizeMib/1024)
	}
	return fmt.Sprintf("%.1f MiB", sizeMib)
}

func FormatDiskSize(bytesCount int64) string {
	if bytesCount == 0 {
		return "0B"
	}
	for _, u := range getFormatDiskUnits() {
		if bytesCount >= u.size {
			value := float64(bytesCount) / float64(u.size)
			if value == float64(int64(value)) {
				return fmt.Sprintf("%d%s", int64(value), u.suffix)
			}
			return fmt.Sprintf("%.1f%s", value, u.suffix)
		}
	}
	return fmt.Sprintf("%dB", bytesCount)
}

// ══════════════════════════════════════════════════════════════════════════════
// Root partition detection
// ══════════════════════════════════════════════════════════════════════════════

// Partition represents a detected partition.
type Partition struct {
	Node   string
	Size   int64 // in sectors
	Type   string
	Name   string
	Label  string
	Fstype string
	Start  int64
}

// PartitionDetector is the interface for partition detectors.
type PartitionDetector interface {
	Name() string
	Weight() float64
	Score(partition Partition, allPartitions []Partition) float64
}

// ── TypeCodeDetector ──

type TypeCodeDetector struct{}

func (TypeCodeDetector) Name() string { return "type_code" }

func (TypeCodeDetector) Weight() float64 {
	return getDetectorWeight("type_code", 0.25)
}

// GPT type GUIDs
const (
	GPTRootX8664   = "44479540-f297-41b2-9af7-d131d5f0458a"
	GPTRootAarch64 = "4f68bce3-e8cd-4db1-96e7-fbcaf984b709"
	GPTESP         = "c12a7328-f81f-11d2-ba4b-00a0c93ec93b"
	GPTSwap        = "0657fd6d-a4ab-43c4-84e5-0933c84b4f4f"
)

// MBR type codes
const (
	MBRLinux    = "83"
	MBREFI      = "ef"
	MBRSwap     = "82"
	MBRExtended = "85"
	MBRLVM      = "8e"
)

func (TypeCodeDetector) Score(partition Partition, allPartitions []Partition) float64 {
	partitionType := strings.ToLower(partition.Type)

	// Root partitions get highest score
	if partitionType == strings.ToLower(GPTRootX8664) ||
		partitionType == strings.ToLower(GPTRootAarch64) {
		return getDetectorScore("root_score", 1.0)
	}

	// Linux MBR type gets medium score
	if partitionType == strings.ToLower(MBRLinux) {
		return getDetectorScore("mbr_linux_score", 0.5)
	}

	// Exclude partitions (ESP, swap, LVM, extended) get negative score
	if partitionType == strings.ToLower(GPTESP) ||
		partitionType == strings.ToLower(GPTSwap) ||
		partitionType == strings.ToLower(MBREFI) ||
		partitionType == strings.ToLower(MBRSwap) ||
		partitionType == strings.ToLower(MBRExtended) ||
		partitionType == strings.ToLower(MBRLVM) {
		return getDetectorScore("exclude_score", -1.0)
	}

	return getDetectorScore("neutral_score", 0.0)
}

// ── LabelDetector ──

type LabelDetector struct{}

func (LabelDetector) Name() string { return "label" }

func (LabelDetector) Weight() float64 {
	return getDetectorWeight("label", 0.25)
}

func (LabelDetector) Score(partition Partition, allPartitions []Partition) float64 {
	label := partition.Name
	if label == "" {
		label = partition.Label
	}
	label = strings.ToLower(label)

	rootIndicators := []string{"root", "cloudimg", "rootfs"}
	for _, indicator := range rootIndicators {
		if strings.Contains(label, indicator) {
			return getDetectorScore("label_root_score", 1.0)
		}
	}

	excludeIndicators := []string{"esp", "efi", "boot", "swap"}
	for _, indicator := range excludeIndicators {
		if strings.Contains(label, indicator) {
			return getDetectorScore("label_exclude_score", -0.5)
		}
	}

	return getDetectorScore("neutral_score", 0.0)
}

// ── SizeDetector ──

type SizeDetector struct{}

func (SizeDetector) Name() string { return "size" }

func (SizeDetector) Weight() float64 {
	return getDetectorWeight("size", 0.25)
}

func (SizeDetector) Score(partition Partition, allPartitions []Partition) float64 {
	sectorBytes := int64(SectorSizeBytes)
	mebibyteBytes := int64(MebibyteBytes)
	sizeMB := float64(partition.Size) * float64(sectorBytes) / float64(mebibyteBytes)

	// Find the largest partition
	maxSizeMB := float64(0)
	for _, p := range allPartitions {
		if p.Size > 0 {
			pSizeMB := float64(p.Size) * float64(sectorBytes) / float64(mebibyteBytes)
			if pSizeMB > maxSizeMB {
				maxSizeMB = pSizeMB
			}
		}
	}

	minRootSizeMB := float64(MinRootSizeMB)
	tooSmallMB := float64(SizeTooSmallMB)

	if sizeMB < tooSmallMB {
		return getDetectorScore("size_too_small_score", -0.5)
	}

	if sizeMB >= minRootSizeMB {
		if sizeMB >= maxSizeMB {
			return getDetectorScore("size_largest_score", 0.5)
		}
		return getDetectorScore("size_root_score", 0.3)
	}

	return getDetectorScore("neutral_score", 0.0)
}

// ── FilesystemDetector ──

type FilesystemDetector struct{}

func (FilesystemDetector) Name() string { return "filesystem" }

func (FilesystemDetector) Weight() float64 {
	return getDetectorWeight("filesystem", 0.25)
}

func (FilesystemDetector) Score(partition Partition, allPartitions []Partition) float64 {
	fstype := strings.ToLower(partition.Fstype)

	rootFilesystems := map[string]bool{"ext4": true, "btrfs": true, "xfs": true, "f2fs": true}
	if rootFilesystems[fstype] {
		return getDetectorScore("filesystem_root_score", 0.5)
	}
	if fstype == "vfat" {
		return getDetectorScore("filesystem_vfat_score", -0.8)
	}
	if fstype == "crypto_luks" || fstype == "" {
		return getDetectorScore("neutral_score", 0.0)
	}

	return getDetectorScore("neutral_score", 0.0)
}

// ── RootPartitionDetector ──

type RootPartitionDetector struct {
	detectors []PartitionDetector
	disabled  map[string]bool
}

func NewRootPartitionDetector(disabledDetectors []string) *RootPartitionDetector {
	d := &RootPartitionDetector{
		detectors: []PartitionDetector{
			TypeCodeDetector{},
			LabelDetector{},
			SizeDetector{},
			FilesystemDetector{},
		},
		disabled: make(map[string]bool),
	}
	for _, name := range disabledDetectors {
		d.disabled[name] = true
	}
	return d
}

func (d *RootPartitionDetector) Register(detector PartitionDetector) {
	d.detectors = append(d.detectors, detector)
}

func (d *RootPartitionDetector) Detect(partitions []Partition) (int, error) {
	if len(partitions) == 0 {
		return 0, errs.RootPartitionDetectionError(nil, "no partitions to evaluate")
	}
	if len(partitions) == 1 {
		return 1, nil
	}

	type scoredPartition struct {
		index int
		score float64
	}

	var scores []scoredPartition
	for i, partition := range partitions {
		total := 0.0
		for _, detector := range d.detectors {
			if d.disabled[detector.Name()] {
				continue
			}
			total += detector.Weight() * detector.Score(partition, partitions)
		}
		scores = append(scores, scoredPartition{index: i + 1, score: total})
	}

	bestScore := math.Inf(-1)
	for _, sp := range scores {
		if sp.score > bestScore {
			bestScore = sp.score
		}
	}

	var bestPartitions []int
	for _, sp := range scores {
		if sp.score == bestScore {
			bestPartitions = append(bestPartitions, sp.index)
		}
	}

	if len(bestPartitions) > 1 {
		tiedParts := make([]string, len(bestPartitions))
		for i, idx := range bestPartitions {
			tiedParts[i] = strconv.Itoa(idx)
		}
		return 0, errs.TieDetectedError(tiedParts, fmt.Sprintf("tie score %f", bestScore), nil)
	}

	if bestScore < 0 {
		return 0, errs.RootPartitionDetectionError(
			partitionsToMaps(partitions),
			fmt.Sprintf("Best score %f < 0, no suitable root partition found", bestScore),
		)
	}

	return bestPartitions[0], nil
}

// partitionsToMaps converts a slice of Partition structs to the []map[string]any
// format expected by RootPartitionDetectionError, matching Python's partition dicts.
func partitionsToMaps(partitions []Partition) []map[string]any {
	result := make([]map[string]any, len(partitions))
	for i, p := range partitions {
		result[i] = map[string]any{
			"node":    p.Node,
			"size":    p.Size,
			"type":    p.Type,
			"name":    p.Name,
			"label":   p.Label,
			"fstype":  p.Fstype,
			"start":   p.Start,
		}
	}
	return result
}

func getDetectorScore(key string, defaultVal float64) float64 {
	if val, ok := DetectorScores[key]; ok {
		return val
	}
	return defaultVal
}

func getDetectorWeight(key string, defaultVal float64) float64 {
	if val, ok := DetectorWeights[key]; ok {
		return val
	}
	return defaultVal
}

File: legacy/python/src/mvmctl/utils/_disk.py
"""Disk size parsing and root partition detection utilities."""

from __future__ import annotations

import logging
import re
from typing import Final, Protocol

from mvmctl import constants
from mvmctl.exceptions import (
    MVMError,
    RootPartitionDetectionError,
    TieDetectedError,
)

logger = logging.getLogger(__name__)

__all__ = [
    "DiskUtils",
    "format_sectors_human_readable",
    "format_disk_size",
    "PartitionDetector",
    "RootPartitionDetector",
    "TypeCodeDetector",
    "LabelDetector",
    "SizeDetector",
    "FilesystemDetector",
]


# ==================== Disk size utilities ====================

# Size multipliers (IEC binary units: KiB, MiB, GiB, TiB)
_SIZE_MULTIPLIERS: Final[dict[str, int]] = {
    "B": 1,
    "K": 1024,
    "KB": 1024,
    "M": 1024**2,
    "MB": 1024**2,
    "G": 1024**3,
    "GB": 1024**3,
    "T": 1024**4,
    "TB": 1024**4,
}

_SIZE_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"^(\d+(?:\.\d+)?)\s*([KMGT]?B?|[kmgt]?b?)?$"
)


class DiskUtils:
    """Disk size parsing and formatting utilities."""

    @staticmethod
    def parse_disk_size_to_bytes(size_str: str) -> int:
        """
        Parse disk size string to bytes.

        Supports: 512M, 1G, 2.5GB, 1024K, etc.
        Case-insensitive. Whitespace allowed between number and unit.

        Args:
            size_str: Size string like "512M", "1G", "2.5GB"

        Returns:
            Size in bytes as integer

        Raises:
            MVMError: If format is invalid

        """
        size_str = size_str.strip().upper()
        match = _SIZE_PATTERN.match(size_str)

        if not match:
            raise MVMError(
                f"Invalid disk size format: '{size_str}'. "
                "Expected format: <number><unit> where unit is B, K, KB, M, MB, G, GB, T, TB"
            )

        number_str, unit = match.groups()
        unit = unit or "B"  # Default to bytes if no unit

        try:
            number = float(number_str)
        except ValueError:
            raise MVMError(f"Invalid number in disk size: '{number_str}'")

        multiplier = _SIZE_MULTIPLIERS.get(unit.upper())
        if multiplier is None:
            raise MVMError(
                f"Unknown size unit: '{unit}'. Valid: B, K, KB, M, MB, G, GB, T, TB"
            )

        bytes_count = int(number * multiplier)

        if bytes_count < 0:
            raise MVMError(f"Disk size cannot be negative: {size_str}")

        return bytes_count


def format_sectors_human_readable(
    size_sectors: int, sector_size: int = 512
) -> str:
    """
    Convert size in sectors to human-readable format (MiB/GiB).

    Args:
        size_sectors: Size in sectors
        sector_size: Sector size in bytes (default 512)

    Returns:
        Human-readable string like "512.0 MiB", "2.5 GiB"

    """
    size_bytes = size_sectors * sector_size
    size_mib = size_bytes / (1024 * 1024)
    if size_mib >= 1024:
        return f"{size_mib / 1024:.1f} GiB"
    return f"{size_mib:.1f} MiB"


def format_disk_size(bytes_count: int) -> str:
    """
    Format bytes to human-readable string.

    Args:
        bytes_count: Size in bytes

    Returns:
        Human-readable string like "1.5G", "512M"

    """
    for unit, multiplier in sorted(
        _SIZE_MULTIPLIERS.items(), key=lambda x: x[1], reverse=True
    ):
        if unit in ("B", "KB", "MB", "GB", "TB"):  # Skip short forms for output
            continue
        if bytes_count >= multiplier:
            value = bytes_count / multiplier
            if value == int(value):
                return f"{int(value)}{unit}"
            return f"{value:.1f}{unit}"
    return f"{bytes_count}B"


# ==================== Root partition detection ====================


class PartitionDetector(Protocol):
    """
    Protocol defining the interface for partition detectors.

    Each detector evaluates partition characteristics and returns a score
    indicating how likely a partition is to be the root filesystem.
    """

    @property
    def name(self) -> str:
        """Detector identifier."""
        ...

    @property
    def weight(self) -> float:
        """Relative weight for this detector in the final score."""
        ...

    def score(
        self,
        partition: dict[str, object],
        all_partitions: list[dict[str, object]],
    ) -> float:
        """
        Evaluate a partition and return a score.

        Args:
            partition: Dictionary containing partition information (e.g., type code,
                label, size, filesystem type).
            all_partitions: List of all detected partitions for context comparison.

        Returns:
            Score between 0.0 and 1.0 indicating partition suitability as root.

        """
        ...


class RootPartitionDetector:
    """
    Detects the most likely root partition using weighted detector heuristics.

    Combines multiple detector strategies (type code, label, size, filesystem)
    to identify the most suitable candidate for use as the root filesystem.
    """

    def __init__(self, disabled_detectors: list[str] | None = None) -> None:
        """
        Initialize the detector with all built-in detectors registered.

        Args:
            disabled_detectors: List of detector names to skip during detection.

        """
        self._detectors: list[PartitionDetector] = [
            TypeCodeDetector(),
            LabelDetector(),
            SizeDetector(),
            FilesystemDetector(),
        ]
        self._disabled = set(disabled_detectors or [])

    def register(self, detector: PartitionDetector) -> None:
        """
        Register a detector for use in root partition detection.

        Args:
            detector: A PartitionDetector implementation to add to the registry.

        """
        self._detectors.append(detector)

    def detect(self, partitions: list[dict[str, object]]) -> int:
        """
        Detect the most likely root partition from a list of candidates.

        Args:
            partitions: List of partition dictionaries to evaluate.

        Returns:
            Index of the detected root partition in the input list.

        Raises:
            RootPartitionDetectionError: If no suitable root partition is found.
            TieDetectedError: If multiple partitions score equally and highest.

        """
        if len(partitions) == 1:
            return 1

        scores: list[tuple[int, float]] = []
        for i, partition in enumerate(partitions):
            total = sum(
                detector.weight * detector.score(partition, partitions)
                for detector in self._detectors
                if detector.name not in self._disabled
            )
            scores.append((i + 1, total))
            logger.debug("Partition %d score: %f", i + 1, total)

        best_score = max(score for _, score in scores)
        best_partitions = [p for p, s in scores if s == best_score]

        if len(best_partitions) > 1:
            raise TieDetectedError(
                [str(p) for p in best_partitions],
                partitions=partitions,
            )

        if best_score < 0:
            raise RootPartitionDetectionError(
                f"Best score {best_score} < 0, no suitable root partition found",
                partitions=partitions,
            )

        return best_partitions[0]


class TypeCodeDetector:
    """
    Detector for identifying root partitions based on partition type codes.

    Linux root partitions typically have specific type codes that indicate
    their purpose in the system.
    """

    # GPT type GUIDs
    GPT_ROOT_X86_64 = "44479540-f297-41b2-9af7-d131d5f0458a"
    GPT_ROOT_AARCH64 = "4f68bce3-e8cd-4db1-96e7-fbcaf984b709"
    GPT_ESP = "c12a7328-f81f-11d2-ba4b-00a0c93ec93b"
    GPT_SWAP = "0657fd6d-a4ab-43c4-84e5-0933c84b4f4f"

    # MBR type codes (as strings from sfdisk JSON)
    MBR_LINUX = "83"
    MBR_EFI = "ef"
    MBR_SWAP = "82"
    MBR_EXTENDED = "85"
    MBR_LVM = "8e"

    @property
    def name(self) -> str:
        return "type_code"

    @property
    def weight(self) -> float:
        return constants.DETECTOR_WEIGHTS.get("type_code", 0.25)

    def score(
        self,
        partition: dict[str, object],
        all_partitions: list[dict[str, object]],
    ) -> float:
        """
        Score a partition based on its type code.

        Args:
            partition: Partition information dictionary.
            all_partitions: All partitions for context. Part of the
                :class:`PartitionDetector` protocol; not used here because
                type-code scoring is absolute (does not depend on neighbors).

        Returns:
            Score based on type code matching root filesystem patterns.

        """
        partition_type = partition.get("type", "")
        if not isinstance(partition_type, str):
            return constants.DETECTOR_SCORES.get("neutral_score", 0.0)

        type_lower = partition_type.lower()

        # Root partitions get highest score
        if type_lower in (
            self.GPT_ROOT_X86_64.lower(),
            self.GPT_ROOT_AARCH64.lower(),
        ):
            return constants.DETECTOR_SCORES.get("root_score", 1.0)

        # Linux MBR type gets medium score
        if type_lower == self.MBR_LINUX.lower():
            return constants.DETECTOR_SCORES.get("mbr_linux_score", 0.5)

        # Exclude partitions (ESP, swap, LVM, extended) get negative score
        if type_lower in (
            self.GPT_ESP.lower(),
            self.GPT_SWAP.lower(),
            self.MBR_EFI.lower(),
            self.MBR_SWAP.lower(),
            self.MBR_EXTENDED.lower(),
            self.MBR_LVM.lower(),
        ):
            return constants.DETECTOR_SCORES.get("exclude_score", -1.0)

        # Unknown types get neutral score
        return constants.DETECTOR_SCORES.get("neutral_score", 0.0)


class LabelDetector:
    """
    Detector for identifying root partitions based on filesystem labels.

    Root partitions often have specific labels like 'ROOT', 'root', or
    variations that indicate their purpose.
    """

    @property
    def name(self) -> str:
        return "label"

    @property
    def weight(self) -> float:
        return constants.DETECTOR_WEIGHTS.get("label", 0.25)

    def score(
        self,
        partition: dict[str, object],
        all_partitions: list[dict[str, object]],
    ) -> float:
        """
        Score a partition based on its filesystem label.

        Args:
            partition: Partition information dictionary.
            all_partitions: All partitions for context. Part of the
                :class:`PartitionDetector` protocol; not used here because
                label scoring is intrinsic (does not depend on neighbors).

        Returns:
            Score based on label matching root filesystem patterns.

        """
        # Get label from partition["name"] or partition["label"], default to empty string
        label = partition.get("name", "") or partition.get("label", "")
        if not isinstance(label, str):
            return constants.DETECTOR_SCORES.get("neutral_score", 0.0)

        label_lower = label.lower()

        # Root indicators - positive score
        root_indicators = ("root", "cloudimg", "rootfs")
        for indicator in root_indicators:
            if indicator in label_lower:
                return constants.DETECTOR_SCORES.get("label_root_score", 1.0)

        # Exclude indicators - negative score
        exclude_indicators = ("esp", "efi", "boot", "swap")
        for indicator in exclude_indicators:
            if indicator in label_lower:
                return constants.DETECTOR_SCORES.get(
                    "label_exclude_score", -0.5
                )

        # No indicators - neutral score
        return constants.DETECTOR_SCORES.get("neutral_score", 0.0)


class SizeDetector:
    """
    Detector for identifying root partitions based on partition size.

    Root filesystems typically have size characteristics that distinguish
    them from small boot/EFI partitions.
    """

    @property
    def name(self) -> str:
        return "size"

    @property
    def weight(self) -> float:
        return constants.DETECTOR_WEIGHTS.get("size", 0.25)

    def score(
        self,
        partition: dict[str, object],
        all_partitions: list[dict[str, object]],
    ) -> float:
        """
        Score a partition based on its size relative to minimum root size.

        Args:
            partition: Partition information dictionary.
            all_partitions: All partitions for context.

        Returns:
            Score based on size being >= MIN_ROOT_SIZE_MB threshold.

        """
        # Get partition size from partition["size"] (in sectors)
        size_value = partition.get("size", 0)
        if not isinstance(size_value, (int, float)):
            return constants.DETECTOR_SCORES.get("neutral_score", 0.0)

        # Convert sectors to MB: size_mb = size_sectors * SECTOR_SIZE / MEBIBYTE
        sector_bytes = constants.CONST_SECTOR_SIZE_BYTES
        mebibyte_bytes = constants.CONST_MEBIBYTE_BYTES
        size_mb = float(size_value) * sector_bytes / mebibyte_bytes

        # Find the largest partition from all_partitions
        max_size_mb = float(0)
        for p in all_partitions:
            p_size = p.get("size", 0)
            if isinstance(p_size, (int, float)):
                p_size_mb = float(p_size) * sector_bytes / mebibyte_bytes
                if p_size_mb > max_size_mb:
                    max_size_mb = p_size_mb

        min_root_size_mb = constants.MIN_ROOT_SIZE_MB
        too_small_mb = constants.SIZE_TOO_SMALL_MB

        # Too small for root filesystem (< too_small_mb)
        if size_mb < too_small_mb:
            return constants.DETECTOR_SCORES.get("size_too_small_score", -0.5)

        # At least MIN_ROOT_SIZE_MB (500MB)
        if size_mb >= min_root_size_mb:
            # Check if this is the largest partition
            if size_mb >= max_size_mb:
                return constants.DETECTOR_SCORES.get("size_largest_score", 0.5)
            else:
                return constants.DETECTOR_SCORES.get("size_root_score", 0.3)

        # Medium-sized partition (too_small_mb - min_root_size_mb, not largest)
        return constants.DETECTOR_SCORES.get("neutral_score", 0.0)


class FilesystemDetector:
    """
    Detector for identifying root partitions based on filesystem type.

    Root filesystems typically use common Linux filesystems like ext4,
    btrfs, or xfs.
    """

    @property
    def name(self) -> str:
        return "filesystem"

    @property
    def weight(self) -> float:
        return constants.DETECTOR_WEIGHTS.get("filesystem", 0.25)

    def score(
        self,
        partition: dict[str, object],
        all_partitions: list[dict[str, object]],
    ) -> float:
        """
        Score a partition based on its filesystem type.

        Args:
            partition: Partition information dictionary.
            all_partitions: All partitions for context. Part of the
                :class:`PartitionDetector` protocol; not used here because
                filesystem-type scoring is intrinsic (does not depend on neighbors).

        Returns:
            Score based on filesystem type matching common root types.

        """
        fstype = partition.get("fstype", "")
        if not isinstance(fstype, str):
            return constants.DETECTOR_SCORES.get("neutral_score", 0.0)

        fstype_lower = fstype.lower()
        root_filesystems = ("ext4", "btrfs", "xfs", "f2fs")

        if fstype_lower in root_filesystems:
            return constants.DETECTOR_SCORES.get("filesystem_root_score", 0.5)
        if fstype_lower == "vfat":
            return constants.DETECTOR_SCORES.get("filesystem_vfat_score", -0.8)
        if fstype_lower in ("crypto_luks", ""):
            return constants.DETECTOR_SCORES.get("neutral_score", 0.0)

        return constants.DETECTOR_SCORES.get("neutral_score", 0.0)
